client/mksubvolume.cc

```diff
@@ -326,45 +378,79 @@ doit()
     // Determine name for new subvolume: It is the target name without the
     // leading filesystem target.
 
-    string subvolume_name = target.substr(fs_target.size() + (fs_target == "/" ? 0 : 1));
+    string subvolume_name = target.substr(fs_target.size() +
+					 (fs_target == "/" || fs_target == target ? 0 : 1));
     if (verbose)
 	cout << "subvolume-name:" << subvolume_name << endl;
 
-    // Execute all steps.
+    // Set up cache for UUID / LABEL resolution in mnt_table_find_source
+    libmnt_cache* cache = mnt_new_cache();
+    libmnt_table* mtab_table = mnt_new_table();
+    mnt_table_set_cache(mtab_table, cache);
+    mnt_table_parse_mtab(mtab_table, NULL);
+
+    libmnt_fs* expected_fs = create_fstab_line(fs, subvol_option, subvolume_name);
+
+    // Consistency checks on (partially) existing entries
+    libmnt_fs* fstab_entry = mnt_table.find_target(target, MNT_ITER_FORWARD);
+    libmnt_fs* mounted_entry = mnt_table_find_target(mtab_table, mnt_fs_get_target(expected_fs),
+						     MNT_ITER_BACKWARD);
+    if (subvolume_name.empty())
+	throw runtime_error("target is a dedicated mountpoint");
+    if (fstab_entry != NULL && strcmp(mnt_fs_get_source(fstab_entry), mnt_fs_get_source(expected_fs)) != 0)
+	throw runtime_error("existing fstab entry doesn't match target device");
+    if (fstab_entry != NULL && strcmp(mnt_fs_get_options(fstab_entry), mnt_fs_get_options(expected_fs)) != 0)
```

Comment:
Even harmless options like "rw" will make this check fail. Are there not options the user might have added that should not stop the program? Please at least add a comment.

Reply:
This was written under the premise that the application will usually be called by scripts with idempotent results from consecutive runs. Manually created or modified entries are deliberately not recognized as "matching", as the result is not the same.

This could be changed, but would significantly increase code complexity (similar to the mounted device detection). Shall I still implement it?

Reply:
Documenting it is also fine for me.

Reply:
I've changed it to check only the `subvol` option (so any other option the user may add will be accepted).

As `subvol` may be given either relative or absolute (`@/something` vs. `/@/something`) a normalization will be done before the comparison.